File: src/java/org/orbeon/oxf/xforms/processor/handlers/xhtml/XXFormsComponentHandler.java
/**
 * Copyright (C) 2010 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.processor.handlers.xhtml;

import org.dom4j.Document;
import org.orbeon.oxf.xforms.StaticStateGlobalOps;
import org.orbeon.oxf.xml.*;
import org.xml.sax.Attributes;
import org.xml.sax.ContentHandler;
import org.xml.sax.Locator;
import org.xml.sax.SAXException;

/**
 * Handle elements to which custom components are bound.
 */
public class XXFormsComponentHandler extends XFormsBaseHandlerXHTML {

    private String elementName;
    private String elementQName;

    public XXFormsComponentHandler() {
        // Don't forward, we must instead feed the handlers with the shadow content
        super(false, false);
    }

    public void start(String uri, String localname, String qName, Attributes attributes) throws SAXException {

        final ElementHandlerController controller = handlerContext.getController();
        final ContentHandler contentHandler = controller.getOutput();

        final String staticId = handlerContext.getId(attributes);
        final String prefixedId = handlerContext.getIdPrefix() + staticId;
        final String effectiveId = handlerContext.getEffectiveId(attributes);

        final StaticStateGlobalOps staticGlobalOps = containingDocument.getStaticOps();

        final String xhtmlPrefix = handlerContext.findXHTMLPrefix();

        this.elementName = staticGlobalOps.getBinding(prefixedId).containerElementName();
        this.elementQName = XMLUtils.buildQName(xhtmlPrefix, elementName);

        // Produce class of the form xbl-foo-bar
        final String classes = "xbl-component xbl-" + qName.replace(':', '-');

        // Start xhtml:span element
        contentHandler.startElement(XMLConstants.XHTML_NAMESPACE_URI, elementName, elementQName, getAttributes(attributes, classes, effectiveId));

        // Push context
        handlerContext.pushComponentContext(prefixedId);

        // Process shadow content if present
        final Document shadowTree = staticGlobalOps.getBinding(prefixedId).fullShadowTree();
        if (shadowTree != null)
            processShadowTree(controller, shadowTree);
    }

    public static void processShadowTree(final ElementHandlerController controller, Document shadowTree) {
        // Tell the controller we are providing a new body
        controller.startBody();

        // Forward shadow content to handler
        // TODO: would be better to handle inclusion and namespaces using XIncludeProcessor facilities instead of custom code
        TransformerUtils.writeDom4j(shadowTree, new EmbeddedDocumentXMLReceiver(controller) {

            private int level = 0;

            public void startElement(String uri, String localname, String qName, Attributes attributes) throws SAXException {

                if (level != 0)
                    super.startElement(uri, localname, qName, attributes);

                level++;
            }

            public void endElement(String uri, String localname, String qName) throws SAXException {

                level--;

                if (level != 0)
                    super.endElement(uri, localname, qName);
            }

            @Override
            public void setDocumentLocator(Locator locator) {
                // NOP for now. In the future, we should push/pop the locator on ElementHandlerController
            }
        });

        // Tell the controller we are done with the new body
        controller.endBody();
    }

    public void end(String uri, String localname, String qName) throws SAXException {

        // Pop context
        handlerContext.popComponentContext();

        final ElementHandlerController controller = handlerContext.getController();
        final ContentHandler contentHandler = controller.getOutput();

        contentHandler.endElement(XMLConstants.XHTML_NAMESPACE_URI, elementName, elementQName);
    }
}

File: src/main/scala/org/orbeon/oxf/xforms/processor/handlers/xhtml/XXFormsComponentHandler.scala
/**
 * Copyright (C) 2007 Orbeon, Inc.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU Lesser General Public License as published by the Free Software Foundation; either version
 * 2.1 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * The full text of the license is available at http://www.gnu.org/copyleft/lesser.html
 */
package org.orbeon.oxf.xforms.processor.handlers.xhtml

import java.lang.StringBuilder

import org.orbeon.oxf.xforms.XFormsConstants.COMPONENT_SEPARATOR
import org.orbeon.oxf.xforms.XFormsUtils._
import org.orbeon.oxf.xforms.analysis.controls.LHHAAnalysis
import org.orbeon.oxf.xforms.control.XFormsControl
import org.orbeon.oxf.xforms.processor.handlers.XFormsBaseHandler.LHHAC
import org.orbeon.oxf.xml._
import org.xml.sax.{Attributes, Locator}

class XXFormsComponentHandler extends XFormsControlLifecyleHandler(false) {

  private var elementName: String = _
  private var elementQName: String = _

  protected override def getContainingElementName = elementName
  protected override def getContainingElementQName = elementQName

  private lazy val binding =
    containingDocument.getStaticOps.getBinding(getPrefixedId) getOrElse (throw new IllegalStateException)

  private lazy val handleLHHA =
    binding.abstractBinding.modeLHHA && ! binding.abstractBinding.modeLHHACustom

  private def hasLabelFor =
    binding.abstractBinding.labelFor.isDefined

  override def init(uri: String, localname: String, qName: String, attributes: Attributes, matched: AnyRef): Unit = {
    super.init(uri, localname, qName, attributes, matched)

    elementName  = binding.abstractBinding.containerElementName
    elementQName = XMLUtils.buildQName(handlerContext.findXHTMLPrefix, elementName)
  }

  protected override def addCustomClasses(classes: StringBuilder, control: XFormsControl): Unit = {
    if (classes.length != 0)
      classes.append(' ')

    classes.append(binding.abstractBinding.cssClasses)
  }

  protected def handleControlStart(
    uri         : String,
    localname   : String,
    qName       : String,
    attributes  : Attributes,
    effectiveId : String,
    control     : XFormsControl
  ): Unit = {

    val prefixedId = getPrefixedId
    val controller = handlerContext.getController

    handlerContext.pushComponentContext(prefixedId)

    // Process shadow content
    XXFormsComponentHandler.processShadowTree(controller, binding.templateTree)
  }

  protected override def handleControlEnd(
    uri         : String,
    localname   : String,
    qName       : String,
    attributes  : Attributes,
    effectiveId : String,
    control     : XFormsControl
  ) = handlerContext.popComponentContext()

  protected override def handleLabel() =
    if (handleLHHA && ! LHHAAnalysis.hasLHHAPlaceholder(containingDocument.getStaticOps.findControlAnalysis(getPrefixedId).get, "label")) {
      if (hasLabelFor) {
        super.handleLabel()
      } else {
          handleLabelHintHelpAlert(
            getStaticLHHA(getPrefixedId, LHHAC.LABEL),
            getEffectiveId,
            getForEffectiveId(getEffectiveId),
            LHHAC.LABEL,
            "span",
            currentControlOrNull,
            isTemplate,
            false
          )
      }
    }

  protected override def handleAlert() = if (handleLHHA) super.handleAlert()
  protected override def handleHint()  = if (handleLHHA) super.handleHint()
  protected override def handleHelp()  = if (handleLHHA) super.handleHelp()

  // If there is a label-for, use that, otherwise don't use @for as we are not pointing to an HTML form control
  // TODO: Most of this should be done statically, not dynamically. See also `findTargetControlFor`.
  override def getForEffectiveId(effectiveId: String) = {

    val labelForStaticIdOpt = binding.abstractBinding.labelFor

    val staticTargetAndLabelForPrefixedIdOpt =
      for {
        labelForStaticId   ← labelForStaticIdOpt
        labelForPrefixedId ← binding.innerScope.prefixedIdForStaticIdOpt(labelForStaticId)
        staticTarget       ← containingDocument.getStaticOps.findControlAnalysis(labelForPrefixedId)
      } yield
        (staticTarget, labelForPrefixedId)

    staticTargetAndLabelForPrefixedIdOpt match {
      case Some((staticTarget, labelForPrefixedId)) ⇒
        // `label-for` is known statically
        for {
          currentControl   ← currentControlOpt // can be missing if we are in template
          targetControlFor ← {
            // Assume the target is within the same repeat iteration
            val suffix              = getEffectiveIdSuffixWithSeparator(currentControl.getEffectiveId)
            val labelForEffectiveId = labelForPrefixedId + suffix

            // Push/pop component context so that handler resolution works
            handlerContext.pushComponentContext(getPrefixedId)
            try XFormsLHHAHandler.findTargetControlForEffectiveId(handlerContext, staticTarget, labelForEffectiveId)
            finally handlerContext.popComponentContext()
          }
        } yield
          targetControlFor
      case None ⇒
        // `label-for` is now known statically, assume it's a nested HTML element
        for {
          labelForStaticId ← labelForStaticIdOpt
          currentControl   ← currentControlOpt // can be missing if we are in template
        } yield
          getRelatedEffectiveId(currentControl.getEffectiveId + COMPONENT_SEPARATOR, labelForStaticId)
    }
  } orNull
}

object XXFormsComponentHandler {

  def processShadowTree(controller: ElementHandlerController, templateTree: SAXStore): Unit = {
    // Tell the controller we are providing a new body
    controller.startBody()

    // Forward shadow content to handler
    // TODO: Handle inclusion/namespaces with XIncludeProcessor instead of custom code.
    templateTree.replay(new EmbeddedDocumentXMLReceiver(controller) {

      var level = 0

      override def startElement(uri: String, localname: String, qName: String, attributes: Attributes): Unit = {

        if (level != 0)
          super.startElement(uri, localname, qName, attributes)

        level += 1
      }

      override def endElement(uri: String, localname: String, qName: String): Unit = {

        level -= 1

        if (level != 0)
          super.endElement(uri, localname, qName)
      }

      override def setDocumentLocator(locator: Locator): Unit = {
        // NOP for now. In the future, we should push/pop the locator on ElementHandlerController
      }
    })

    // Tell the controller we are done with the new body
    controller.endBody()
  }
}